File: solutions/Combination-Sum-II/Combination-Sum-II.cpp

class Solution {
public:
    void dfs(vector<vector<int> > &ans,vector<int> vec,vector<int> num,int target,int pos)
    {
        if(target==0)
        {
            ans.push_back(vec);
            return ;
        }
        int last=-1;
        for(int i=pos;i<num.size();i++)
        {
            if(num[i]==last)
                continue;
            if(target<num[i])
                return;
            last=num[i];
            vec.push_back(num[i]);
            dfs(ans,vec,num,target-num[i],i+1);
            vec.pop_back();
        }
    }
    vector<vector<int> > combinationSum2(vector<int> &num, int target) {
        sort(num.begin(),num.end());
        vector<vector<int> > ans;
        vector<int> vec;
        dfs(ans,vec,num,target,0);
        return ans;
    }
};


 

File: solutions/Combination-Sum-II/Combination-Sum-II.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum2 = function (candidates, target) {
  candidates.sort((x, y) => x < y ? -1 : 1);
  const ans = [];
  const dfs = (tar, vec, pos) => {
    if (tar === 0) {
      ans.push(vec.slice());
      return;
    }
    let last = -1;
    for (let i = pos; i < candidates.length; i++) {
      if (candidates[i] === last) {
        continue;
      }
      if (tar < candidates[i]) {
        return;
      }
      last = candidates[i];
      vec.push(candidates[i]);
      dfs(tar - candidates[i], vec, i + 1);
      vec.pop();
    }
  };
  dfs(target, [], 0);
  return ans;
};